src/routes/api/client-key/recover-client-key.ts

```diff
@@ -0,0 +1,94 @@
+import { Request, Response } from "express";
+import { Installation } from "models/installation";
+import { getHashedKey } from "models/sequelize";
+import { Op } from "sequelize";
+import safeJsonStringify from "safe-json-stringify";
+import axios from "axios";
+
+const DEFAULT_BATCH_SIZE = 5000;
+
+export const RecoverClientKeyPost = async (req: Request, res: Response): Promise<void> => {
+
+	const startInstallationId = Number(req.query.startInstallationId) || 0;
+	const batchSize = Number(req.query.batchSize) || DEFAULT_BATCH_SIZE;
+	const overrideExisting = (Number(req.query.overrideExisting) || 0) === 1;
+
+	res.status(200);
+
+	res.write(`Start recovering client keys for installation starting from ${startInstallationId}\n`);
+
+	const foundInstallations: Installation[] = await Installation.findAll({
+		limit: batchSize,
+		where: {
+			... overrideExisting ? undefined : { jiraClientKey: null },
+			"id": {
+				[Op.gte]: startInstallationId
+			}
+		}
+	});
+
+	res.write(`Found ${foundInstallations.length} installations to process.\n`);
+
+	let successCount = 0;
+	let failCount = 0;
+
+	for (const installation of foundInstallations) {
+
+		try {
+
+			const originClientKey = await getAndVerifyOriginClientKey(installation);
+			installation.jiraClientKey = originClientKey;
+			await installation.save();
+
+			successCount++;
+		} catch (e) {
+			failCount++;
+			res.write(`Something wrong when processing installation id: ${installation.id}: ${safeJsonStringify(e)}\n`);
+		}
+
+	}
+
+	foundInstallations.sort((a,b)=>a.id - b.id);
+	const lastId = foundInstallations.length === 0 ? NaN : foundInstallations[foundInstallations.length - 1].id;
+
+	res.write(`All done, SUCCESS: ${successCount} - FAILED: ${failCount}, lastId: ${lastId}`);
+
+	res.end();
+
+};
+
+const getAndVerifyOriginClientKey = async ({ id, jiraHost, clientKey: hashedClientKeyInDB }: Installation): Promise<string>  => {
+
+	if (!jiraHost) {
+		throw { msg: `Cannot find jiraHost`, id };
+	}
+	if (!jiraHost.endsWith("/")) jiraHost = jiraHost + "/";
+
+	const result = await axios(`${jiraHost}plugins/servlet/oauth/consumer-info`);
```

Comment:
Make sure you replace this with outbound proxy!!!
example https://github.com/atlassian/github-for-jira/pull/1690/files

Reply:
Actually I replace it with whitelist proxy (so that we can hit the jira-dev.com). Let me know if that's okay.
--> whitelist proxy should be a "super-set" of outbound proxy I believe.